File: app-ios/Packages/module-ai/Sources/ModuleAI/Application/UseCases/AIChat/LoadChartDataUseCase.swift
//
//  LoadChartDataUseCase.swift
//  ModuleAI
//
//  Created by Codex on 2026/1/8.
//

import Foundation

struct LoadChartDataUseCase {
    private let repository: AIChatRepository

    init(repository: AIChatRepository) {
        self.repository = repository
    }

    func execute(
        name: FunctionName,
        historyId: Int,
        arguments: FunctionArguments
    ) async throws -> UseCaseResult<LoadChartDataOutput> {
        guard name.argumentKind == arguments.argumentKind else {
            return .failure(.message("函数参数类型不匹配。"))
        }
        
        let model = try await repository.loadChartData(
            name: name,
            historyId: historyId,
            arguments: arguments
        )
        let payload = AIChatApplicationMapper.makeChartPayload(from: model)
        guard let payload, payload.series.isEmpty == false else {
            return .failure(.message(
                payload?.emptyMessage
                ?? "数据分析还在测试阶段，很快就能上线，敬请期待！"
            ))
        }

        return .success(LoadChartDataOutput(payload: payload))
    }
}

File: app-android/feature/ai-chat/src/main/java/com/aidatainsight/android/feature/aichat/application/usecase/LoadChartDataUseCase.kt
package com.aidatainsight.android.feature.aichat.application.usecase

import com.aidatainsight.android.core.model.contract.FunctionArguments
import com.aidatainsight.android.core.model.contract.FunctionName
import com.aidatainsight.android.feature.aichat.application.AIChatApplicationMapper
import com.aidatainsight.android.feature.aichat.application.model.LoadChartDataOutput
import com.aidatainsight.android.feature.aichat.application.model.UseCaseResult
import com.aidatainsight.android.feature.aichat.domain.AIChatRepository

class LoadChartDataUseCase(
    private val repository: AIChatRepository,
) {
    suspend operator fun invoke(
        name: FunctionName,
        historyId: Int,
        arguments: FunctionArguments,
    ): UseCaseResult<LoadChartDataOutput> {
        if (name.argumentKind != arguments.kind) {
            return UseCaseResult.Failure("函数参数类型不匹配。")
        }

        val model = repository.loadChartData(
            name = name,
            historyId = historyId,
            arguments = arguments,
        )
        val payload = AIChatApplicationMapper.makeChartPayload(model)
        if (payload == null || payload.series.isEmpty()) {
            return UseCaseResult.Failure(
                payload?.emptyMessage ?: AIChatApplicationMapper.chartFallbackMessage(),
            )
        }
        return UseCaseResult.Success(LoadChartDataOutput(payload = payload))
    }
}
